src/libcore/array.rs

```diff
@@ -406,6 +408,73 @@ mod impls_using_const_generics {
         }
     }
 
+    #[unstable(feature = "array_value_iter", issue = "0")]
+    impl<T, const N: usize> IntoIterator for [T; N]
+    where
+        [T; N]: LengthAtMost32,
+    {
+        type Item = T;
+        type IntoIter = IntoIter<T, {N}>;
+
+        fn into_iter(self) -> Self::IntoIter {
+            IntoIter {
+                data: ManuallyDrop::new(self),
+                pos: 0,
+            }
+        }
+    }
+
+    /// A by-value array iterator.
+    ///
+    /// This struct is created by the `IntoIterator::into_iter`] method on
+    /// arrays.
+    #[unstable(feature = "array_value_iter", issue = "0")]
+    pub struct IntoIter<T, const N: usize> {
+        /// This is the array we are iterating over. Elements that were already
+        /// yielded (the ones with index `< self.pos`) have been read via
+        /// `ptr::read` and must not be accessed anymore!
+        data: ManuallyDrop<[T; N]>,
+        pos: usize,
+    }
+
+    #[unstable(feature = "array_value_iter", issue = "0")]
+    impl<T, const N: usize> Iterator for IntoIter<T, {N}> {
+        type Item = T;
+        fn next(&mut self) -> Option<Self::Item> {
+            if self.pos >= N {
+                return None;
+            }
+
+            let idx = self.pos;
+            self.pos += 1;
+            Some(unsafe { ptr::read(self.data.get_unchecked_mut(idx)) })
+        }
+    }
+
+    #[unstable(feature = "array_value_iter", issue = "0")]
+    impl<T, const N: usize> Drop for IntoIter<T, {N}> {
+        fn drop(&mut self) {
+            // Only drop the elements that were not yielded yet.
+            unsafe {
+                ptr::drop_in_place(&mut self.data[self.pos..]);
+            }
+        }
+    }
+
+    #[unstable(feature = "array_value_iter", issue = "0")]
+    impl<T: fmt::Debug, const N: usize> fmt::Debug for IntoIter<T, {N}> {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            // Only print the elements that were not yielded yet: we cannot
+            // access the yielded elements anymore. This might be a bit
+            // confusing, though, so maybe a `[_, _, _, 1, 2,]` formatting or
```

Comment:
Keep it (almost) as-is (and no need to output `pos` either) since this is consistent with `slice::Iter`:
```rust
fn main() {
    let vec = (0..10).collect::<Vec<_>>();
    let mut iter = vec.iter();
    let _ = iter.next();
    let _ = iter.next();
    dbg!(iter);
}
```
yields:
```rust

[src/main.rs:6] iter = Iter(
    [
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
    ],
)
```
with definition: https://github.com/rust-lang/rust/blob/03f19f7ff128a3b01eeab3f87f04cce22883f006/src/libcore/slice/mod.rs#L3327-L3342